annotatedcsv/decode.go

```diff
@@ -0,0 +1,463 @@
+// It's called once per section.
+func (r *Reader) initColumns(t reflect.Type, columns []Column) error {
+	if t == r.decodeType {
+		return nil
+	}
+	if t.Kind() != reflect.Ptr {
+		return fmt.Errorf("cannot decode into non-pointer type %v", t)
+	}
+	et := t.Elem()
+	if et.Kind() != reflect.Struct && et.Kind() != reflect.Slice {
+		return fmt.Errorf("cannot decode into %v", t)
+	}
+	setters := make([]fieldSetter, len(columns))
+	switch et.Kind() {
+	case reflect.Struct:
+		fieldsMap := make(map[string]reflect.StructField)
+		fields := ireflect.VisibleFields(et)
+		for _, f := range fields {
+			name := f.Name
+			if tag, ok := f.Tag.Lookup("csv"); ok {
+				switch {
+				case tag == "-":
+					continue
+				case tag == "-,":
+					tag = "-"
+				case strings.Contains(tag, ","):
+					i := strings.Index(tag, ",")
+					if i == len(tag)-1 {
+						tag = tag[:len(tag)-1]
+					} else {
+						return fmt.Errorf("tag attributes are not supported")
+					}
+				}
+				name = tag
+			}
+			fieldsMap[name] = f
+		}
+
+		for i, col := range columns {
+			f, ok := fieldsMap[col.Name]
+			if !ok {
+				continue
+			}
+			ftype := f.Type
+			colType, ok := columnTypes[col.Type]
+			if !ok {
+				// ignore invalid type and use string
+				colType = stringCol
+			}
+			fkind := fieldKindOf(ftype)
+			convert, ok := conversions[conv{colType, fkind}]
+			if !ok {
+				return fmt.Errorf("cannot convert from column type %s to %v", col.Type, ftype)
+			}
+			setters[i] = func(v reflect.Value, colIndex int) error {
+				if err := r.convertColumnValue(v.FieldByIndex(f.Index), colIndex, convert); err != nil {
+					return err
+				}
+				return nil
+			}
+		}
+		r.decodeRow = func(v reflect.Value, row []string) error {
+			v = v.Elem()
+			for i, c := range r.colSetters {
+				if c == nil {
+					continue
+				}
+				if err := c(v, i); err != nil {
+					return err
+				}
+			}
+			return nil
+		}
+	case reflect.Slice:
+		s := et.Elem()
+		if s != stringType && s != emptyInterfaceType {
+			return fmt.Errorf("cannot decode into pointer to slice of %v", s)
+		}
+		fkind := fieldKindOf(s)
+		for i, col := range columns {
+			colType, ok := columnTypes[col.Type]
+			if !ok {
+				// ignore invalid type and use string
+				colType = stringCol
+			}
+
+			convert, ok := conversions[conv{colType, fkind}]
+			if !ok {
+				return fmt.Errorf("cannot convert from column type %s to %v", col.Type, s)
+			}
+			setters[i] = func(v reflect.Value, colIndex int) error {
+				if err := r.convertColumnValue(v.Index(colIndex), colIndex, convert); err != nil {
+					return err
+				}
+				return nil
+			}
+		}
+		r.decodeRow = func(v reflect.Value, row []string) error {
+			e := v.Elem()
+			c := len(r.colSetters)
+			if e.Cap() < c {
+				e = reflect.MakeSlice(e.Type(), c, c)
+			} else {
+				e = e.Slice(0, c)
+			}
+			for i := 0; i < c; i++ {
+				if err := r.colSetters[i](e, i); err != nil {
+					return err
+				}
+			}
```

Comment:
Maybe slightly more efficient if the compiler isn't that clever:

```suggestion
			for _, setf := range r.colSetters {
			    if err := setf(e, i); err != nil {
			        return err
			    }
			}
```